src/dominators.rs

```diff
@@ -0,0 +1,277 @@
+//! # The Dominance Relation
+//!
+//! In a directed graph with a root node `R`, a node `A` is said to *dominate* a
+//! node `B` iff every path from `R` to `B` contains `A`.
+//!
+//! The node `A` is daid to *strictly dominate the node `B` iff `A` dominates
+//! `B` and `A != B`.
+//!
+//! The node `A` is said to be the *immediate dominator* of a node `B` iff it
+//! strictly dominates `B` and does not dominate any other nodes which also
+//! dominate `B` in turn.
+
+use std::collections::{HashMap, HashSet};
+use std::hash::Hash;
+
+use visit::{DfsPostOrder, GraphBase, IntoNeighbors, Visitable, Walker};
+
+/// The dominance relation for some graph and root.
+#[derive(Debug, Clone)]
+pub struct Dominators<G>
+    where G: GraphBase,
+          G::NodeId: Eq + Hash
+{
+    root: G::NodeId,
+    dominators: HashMap<G::NodeId, G::NodeId>,
+}
+
+impl<G> Dominators<G>
+    where G: GraphBase,
+          G::NodeId: Eq + Hash
+{
+    /// Get the root node used to construct these dominance relations.
+    pub fn root(&self) -> G::NodeId {
+        self.root
+    }
+
+    /// Get the immediate dominator of the given node.
+    ///
+    /// Returns `None` for any node that is not reachable from the root, and for
+    /// the root itself.
+    pub fn immediate_dominator(&self, node: G::NodeId) -> Option<G::NodeId> {
+        if node == self.root {
+            None
+        } else {
+            self.dominators.get(&node).cloned()
+        }
+    }
+
+    /// Iterate over the given node's that strict dominators.
+    ///
+    /// If the given node is not reachable from the root, then `None` is
+    /// returned.
+    pub fn strict_dominators(&self, node: G::NodeId) -> Option<Iter<G>> {
+        if self.dominators.contains_key(&node) {
+            Some(Iter {
+                dominators: self,
+                node: self.immediate_dominator(node),
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Iterate over all of the given node's dominators (including the given
+    /// node itself).
+    ///
+    /// If the given node is not reachable from the root, then `None` is
+    /// returned.
+    pub fn dominators(&self, node: G::NodeId) -> Option<Iter<G>> {
+        if self.dominators.contains_key(&node) {
+            Some(Iter {
+                dominators: self,
+                node: Some(node),
+            })
+        } else {
+            None
+        }
+    }
+}
+
+/// Iterator for a node's dominators.
+pub struct Iter<'a, G>
```

Comment:
A good rule for iterators is that they are named after the function that created them (You'll find this is true of the `Iter` you find in libstd!). This one looks like it should be something closer to `Dominators`.